src/substrate-core-modifications.md

```diff
@@ -0,0 +1,174 @@
+    - Adds `IpfsRequest`, `IpfsRequestId`, `IpfsRequestStatus`, `IpfsResponse`
+    - Defines `IpfsPendingRequest`
+    - Adds pending `ipfs_requests` and `expected_ipfs_requests` to `OffchainState`
+    - Defines its own `ipfs_request_start` and `ipfs_response_wait`
+- `primitives/core/src/offchain/mod.rs`
+    - Defines `IpfsRequest` and `IpfsResponse` enums
+    - What is IPFS = 254, Capability, ok.. what does it mean to have 255?
+    - Defines its own `ipfs_request_start` and `ipfs_response_wait`
+        - On Externalities trait of Box<T> and LimitedExternalities
+        - uses `Capability::Ipfs`
+
+#### Runtime
+
+- `primitives/runtime/src/offchain/ipfs.rs`
+    - high-level helpers for making IPFS requests from Offchain Workers.
+    - Defines `PendingRequest` and `Response` and how the former becomes the latter (or an Error)
+
+#### I/O
+
+- `primitives/io/src/lib.rs`
+    - Adds `IpfsRequest`, `IpfsRequestId`
+    - Defines the `ipfs_request_start`, and `ipfs_response_wait` functions
+
+### Client
+
+- `client/service/src/builder.rs`
+    - Adds a new async runtime here for ipfs to both the full and light clients
+- `client/service/src/task_manager/mod.rs`
+    - adds public `ipfs_rt` field to the task manager
+
+#### Offchain Client
+
+Lots of stuff here too
+
+- `client/offchain/Cargo.toml`
+    - Adds tokio
+- `client/offchain/src/api/ipfs.rs`
+    - `IpfsApi` and `IpfsWorker`
+    - Big part of the implementation here
+- `client/offchain/src/api.rs`
+    - uses `IpfsRequest`, `IpfsRequestId`, and `IpfsRequestStatus`
+    - defines `ipfs_request_start`
+    - Includes ipfs
+    - Runs the node API
+- `client/offchain/src/lib.rs`
+
+### Binaries
+
+#### `node`
+
+- `bin/node/cli/src/service.rs`
+    - integrates `task_manager.ipfs_rt` from client
+
+#### `node-template`
+
+- `bin/node-template/node/src/service.rs`
+    - integrates `task_manager.ipfs_rt` from client
+
+### Types
+
+#### IpfsRequestId (pub u16)
+
+Type alias for `pub u16`
+
+#### IpfsError
+
+- `DeadlineReached = 1` - The requested action couldn't been completed within a deadline.
+- `IoError = 2` - There was an IO Error while processing the request.
+- `Invalid = 3` - The ID of the request is invalid in this context.
+
+#### IpfsRequestStatus
+pf
+/// Deadline was reached while we waited for this request to finish.
+	///
+	/// Note the deadline is controlled by the calling part, it not necessarily
+	/// means that the request has timed out.
+	DeadlineReached,
+	/// An error has occurred during the request, for example a timeout or the
+	/// remote has closed our socket.
+	///
+	/// The request is now considered destroyed. To retry the request you need
+	/// to construct it again.
+	IoError(Vec<u8>),
+	/// The passed ID is invalid in this context.
+- `Invalid`
+	/// The request has finished successfully.
+- `Finished(IpfsResponse)`
+
+#### `IpfsRequest`
+
+An enum of request types. Note that some of these are not yet implemented as
```

Comment:
```suggestion
An enum of request types. Note that some of these are not implemented as
```
we don't necessarily need to add them in the future ourselves, but nothing is stopping the users from utilizing those